### PR TITLE
[CPP Graph] ChatGLM-2 Enabling

### PR DESCRIPTION
**Pull request by Zhenzhong1**
## Type of Change

Feature add

## Description

New models add:
* [ChatGLM-2](https://huggingface.co/THUDM/chatglm2-6b)

## Expected Behavior & Potential Risk

N/A.

## How has this PR been tested?

Manual and CI tests.

## Dependency Change?

N/A.

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/application/chat/main_chat.cpp [hunk 1]
```diff
@@ -68,15 +82,17 @@ int main(int argc, char** argv) {
...
   }
+
   model_archs mt = model_name_to_arch::init().find(params.model_name);
-  if (mt == MODEL_UNKNOWN) {
-    fprintf(stderr, "error, please set model_name \n");
-    exit(0);
-  }
+  // if (mt == MODEL_UNKNOWN) {
+  //   fprintf(stderr, "error, please set model_name \n");
+  //   exit(0);
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/application/chat/main_chat.cpp:94]: remove unuse code

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h [hunk 2]
```diff
@@ -64,7 +64,7 @@
 #ifdef __cplusplus
 extern "C" {
 #endif
-enum model_archs { MODEL_UNKNOWN, MODEL_LLAMA, MODEL_GPTJ, MODEL_MPT, MODEL_GPTNEOX, MODEL_STARCODER, MODEL_FALCON };
+enum model_archs { MODEL_UNKNOWN, MODEL_LLAMA, MODEL_GPTJ, MODEL_MPT, MODEL_GPTNEOX, MODEL_STARCODER, MODEL_FALCON, MODEL_CHATGLM, MODEL_CHATGLM2, MODEL_CHATGLM1};
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:67]: why there are three model rathe than MODEL_CHATGLM2, MODEL_CHATGLM1 ？ 
what MODEL_CHATGLM stand for ？

File: intel_extension_for_transformers/llm/runtime/graph/application/chat/main_chat.cpp (hunk 1 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/application/chat/main_chat.cpp]: fixed

**Review by a32543254 — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py [hunk 3]
```diff
@@ -94,6 +94,16 @@ def main(args_in: Optional[List[str]] = None) -> None:
     fout.write(struct.pack("i", 0))
     fout.write(struct.pack("i", 0))  # word_embed_proj_dim (for opt)
     fout.write(struct.pack("i", 0))  # do_layer_norm_before (for opt)
+
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py:100]: please also add change to convert_bloom.py

Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py:100]: Done

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp [hunk 4]
```diff
@@ -196,7 +210,17 @@ int main(int argc, char** argv) {
...
   }
-  auto embd_inp = ::model_tokenize(ctx, params.prompt, add_bos);
+  
+  std::vector<int> embd_inp;
+  if (params.model_arch == MODEL_CHATGLM2 || params.model_arch == MODEL_CHATGLM1) {
+      std::vector<std::string> prompts;
+      prompts.push_back(params.prompt);
+      std::string prompt = build_prompt(prompts);
+      embd_inp = ::model_tokenize(ctx, prompt, false);
+      embd_inp.insert(embd_inp.begin(), {64790, 64792}); // special prefix
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp:220]: what are `64790` and `64792` for? `bos_id`?

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c [hunk 5]
```diff
@@ -2778,7 +2778,7 @@ struct ne_tensor* ne_soft_max_inplace(struct ne_context* ctx, struct ne_tensor*
 // ne_rope
 
 struct ne_tensor* ne_rope_impl(struct ne_context* ctx, struct ne_tensor* a, int n_past, int n_dims, int mode,
-                               bool inplace) {
+                               int n_ctx, bool inplace) {
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:2781]: I don't know if it is a good idea to add an extra input parameter `n_ctx` here. BTW, `n_ctx` can be get by `ctx->model.hparams.n_ctx`

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+      ne_tensor *attn_probs = ne_soft_max_inplace(ctx0, attn_scores); // [heads, qlen, klen]
+
+      ne_tensor *context_layer = ne_mul_mat(ctx0, value_layer, attn_probs); // [heads, qlen, head_size]
+      
+      context_layer = ne_cont(ctx0, ne_permute(ctx0, context_layer, 0, 2, 1, 3));
+      
+      context_layer = ne_reshape_2d(ctx0, context_layer, hidden_size, qlen);
+
+      // struct ne_tensor *attn_output = dense.forward(ctx0, context_layer);
+      // Linear如果有bias就需要add，如果没有就不需要。到底有没有bias呢？？？？？？
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:287]: it depends on model arch. del this annotation？

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+      cur = ne_mul_mat(ctx0, model.layers[il].attn[2], context_layer);
+      cur = ne_add(ctx0, cur, model.layers[il].attn[3]);
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:289]: no need repeat when add bias?

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+//     int err = model_apply_lora_from_file(lctx, params.lora_adapter.c_str(),
+//                                          params.lora_base.empty() ? NULL : params.lora_base.c_str(), params.n_threads);
+//     if (err != 0) {
+//       fprintf(stderr, "%s: error: failed to apply lora adapter\n", __func__);
+//       return NULL;
+//     }
+//   }
+
+//   return lctx;
+// }
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:487]: maybe need to clean some useless annotations in this file

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 9]
```diff
@@ -211,9 +211,20 @@ struct model_file_loader {
     file.read_raw(&hparams.alibi_bias_max, sizeof(float));
     file.read_raw(&hparams.clip_qkv, sizeof(float));
     hparams.par_res = file.read_u32();
+
     hparams.word_embed_proj_dim = file.read_u32();
     hparams.do_layer_norm_before = bool(file.read_u32());
+    
+    // For ChatGLM-1 & 2
+    hparams.bos_token_id = file.read_u32();
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:219]: As we discussed before, I think the special tokens like `eos_id`, and `pad_id` should be members of `vocab`. line 230 for reference.  And `vocab` already has `eos_id` and `bos_id`

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 10]
```diff
@@ -211,9 +211,20 @@ struct model_file_loader {
...
+    hparams.eos_token_id = file.read_u32();
+    hparams.pad_token_id = file.read_u32();
+    hparams.sep_token_id = file.read_u32();
+    hparams.multi_query_group_num = file.read_u32();
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:223]: Is it for `MQA (multi-query-attention)`?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 11]
```diff
@@ -211,9 +211,20 @@ struct model_file_loader {
...
+    hparams.ffn_hidden_size = file.read_u32();
+    hparams.inner_hidden_size = file.read_u32();
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:225]: Is it == `n_embed`?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp (hunk 4 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_run.cpp:220]: These are special prefix. The tokenizer of chatglm-2 must add them as special tag.

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 5 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:2781]: The ggml rope has two speical branch for gptjneox and chatglm1.
https://github.com/ggerganov/ggml/blob/master/src/ggml.c#L12781

We use the ggml rope kernel. maybe we should keep the same as ggml. don't know the reason why delete the glm branch before.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp (hunk 7 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:289]: not sure. @yuchengliu1  does the kernel of add support broadcast now?

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp (hunk 8 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:487]: Yes. I will do cleancode in these files of chatglm-1.

This PR is for the chatglm-2 enabling.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 9 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:219]: I will modify them in the PR of the chatglm-1 enabling.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp (hunk 6 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:287]: Yes, chatglm-1 and 2 are different. I will do cleancode first.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp (hunk 7 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp:289]: But I will check them in the next PR of the chatglm-1 enabling.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 11 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:225]: what does it mean?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 10 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:223]: Yes

**Review by zhenwei-intel — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h [hunk 12]
```diff
@@ -137,6 +152,9 @@ struct model_layer {
 
   // ff
   struct ne_tensor* ffn[MODEL_MAX_FFN];
+
+  struct ne_tensor* k_cache;
+  struct ne_tensor* v_cache;
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:157]: Does it duplicated with `model_struct::kv_self`?

Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:157]: It works for ChatGLM-2. I will double check these parameters later for the ChatGLM-1 enabling.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 10 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:264]: we already have `n_head_kv` param for `MQA`. Is it the same?

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 11 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:266]: I saw you set `0` for `inner_hidden_size`.

**Comment by airMeng** [comment]
where is performance data and extension tests?

**Comment by Zhenzhong1** [comment]
> where is performance data and extension tests?

I will update later.

**Comment by airMeng** [comment]
> > where is performance data and extension tests?
> 
> I will update later.

revert in https://github.com/intel/intel-extension-for-transformers/pull/266

**Comment by airMeng** [comment]
and also solve all conversations

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 11 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:266]: ChatGLM-2 requires theses:
    hparams.multi_query_group_num = file.read_u32();
    hparams.ffn_hidden_size = file.read_u32();

ChatGLM-1 requires hparams.inner_hidden_size.

I will double check whether inner_hidden_size is the same as n_embed.

File: intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp (hunk 6 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/chatglm/chatglm1.cpp]: fixed

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h (hunk 10 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:264]: I will check later. MQA has been recored in the TODO of this PR.

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 5 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:2932]: The rope requires the size of prompts but not the fixed model.haprams.n_ctx.

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h (hunk 12 above)
Comment (Zhenzhong1) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_types.h:157]: Not duplicated. The normal KV cache for all layers are stacked in a tensor. But chatglm-2 is that each layer has its own the KV cache.

**Pull request merged**